### ISSUE TITLE
An annotation for true read-only constants

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#10944]
There should be a way to define true constants in the read-only sections of compiled programs. The current convention is to use an `Array`, `StaticArray`, or sometimes `String` for large constant tables. This has a few drawbacks:

* #2485
* The constants aren't really immutable;
* There is still a runtime start-up cost associated with building large `Array` or `Hash` constants;
* Accidentally passing a `StaticArray` by value produces very bad LLVM IR.

I propose a `@[ReadOnly]` annotation that can be applied to constant definitions; the constant initializer must be "reasonably simple", and its contents are placed into read-only memory. Later we could also add ways to detect writes to those read-only constants during compilation, e.g. forbidding method calls on them that assign to instance variables.

The code generator already has the infrastructure for this, since `String` constants are supported after all. What's difficult is specifying the "reasonably simple" constant expressions allowed on the right-hand sides of constant definitions. The following should be good enough for a first iteration:

* `NumberLiteral`s, `BoolLiteral`s, `CharLiteral`s, and `StringLiteral`s are constant expressions.
* A `TupleLiteral` or `NamedTupleLiteral` is a constant expression if all of its elements or values are constant expressions.
* If `T` resolves to a non-union primitive number type, and `exp1`, `exp2` etc. are all constant expressions, then `T.slice(exp1, exp2, ...)` and `T.static_array(exp1, exp2, ...)` are constant expressions.
* If `T` resolves to a `Slice` or `StaticArray` type (uninstantiated or constant type instance), and `exp1`, `exp2` etc. are all constant expressions, then `T[exp1, exp2, ...]` is a constant expression.
* Unions are not constant expressions.

Constant types, i.e. the possible types of read-only constants, are defined as below:

* All primitive number types, `Bool`, `Char`, and `String` are constant types.
* `Tuple(*T)` or `NamedTuple(**T)` is a constant type if all of `*T` or `**T` are constant types.
* `StaticArray(T, N)` or `Slice(T)` is a constant type if `T` is a constant type.
* Irreducible union types are not constant types.

Possible extensions are arbitrary `Value`s, arbitrary `Reference`s without inner pointers, and finally arbitrary objects. Unfortunately, `Array` and `Hash` belong to the last category, so associative lookup tables will need alternate encodings even with such an annotation present.

Some complex constant initializers can be moved into the macro language. A notable example is `String::CHAR_TO_DIGIT`:

```crystal
{% begin %}
  @[ReadOnly]
  CHAR_TO_DIGIT = Int8.static_array({% for i in 0..0xFF %}
    {% if i >= 0x30 && i <= 0x39 %}    {{ i - 0x30 }},
    {% elsif i >= 0x41 && i <= 0x5a %} {{ i - 0x41 + 10 }},
    {% elsif i >= 0x61 && i <= 0x7a %} {{ i - 0x61 + 10 }},
    {% else %}                         -1,
    {% end %}
  {% end %})
{% end %}

CHAR_TO_DIGIT['3'.ord.to_u8] # => 3
CHAR_TO_DIGIT['g'.ord.to_u8] # => 16
CHAR_TO_DIGIT[0] = 0         # Invalid memory access (signal 11)
```

Continued from https://github.com/crystal-lang/crystal/issues/4516#issuecomment-879508619.

**Comment by straight-shoota** [comment]
Sounds good :+1:

Just wanted to note that primitive types technically don't need an annotation because they're already immutable. You can't change a number constant, not even through pointer assignment.

**Comment by HertzDevil** [comment]
`StaticArray` is a struct, so `data = DATA` should create a copy of the constant by _value_, and mutations on `data` therefore should be allowed; it is only `DATA.[]=` that should be prohibited here.

Likewise, a `Slice` generated from a `[@ReadOnly]` constant is itself immutable; `Slice`s are copied by value, and no methods on `Slice` modify self's `@pointer`, `@size`, or `@read_only`. In this case modifications to the underlying constant data are protected during runtime by `Slice#read_only?`.

As mentioned in the original issue, a good indicator of a mutating def is any def that writes to an instance variable, calls a mutating method on an instance variable, or recursively contains a mutating call on self, so `Point#x=` and `Point#y=` are mutators. But if this is not enough, we could introduce another annotation:

```crystal
struct StaticArray(T, N)
  # does not assign to `@buffer`
  # does not call a mutating method on `@buffer`
  # does not recursively call a mutating method with `self` as receiver
  # therefore, `#to_unsafe` should not be considered mutating
  def to_unsafe
    pointerof(@buffer)
  end

  # mutates self despite otherwise being considered non-mutating
  @[Mutates]
  def []=(index : Int, value : T)
    to_unsafe[index] = value
  end
end
```

IMO having these 2 annotations is easier to learn and migrate existing code to than adding a new built-in type.

**Comment by Fryguy** [comment]
Flipping the idea around, is there ever a time when an author would set a constant and _want_ it to be mutable?  If no, then do we even need the annotation?  I think the fact that it's a constant is enough for us to know that the user expects it to be immutable, and therefore the compiler can put it into ReadOnly memory.  I like the proposal of a subset being good enough for a first iteration. (I think I'm saying the same thing that asterite said 😄)

**Comment by asterite** [comment]
Good points. Yeah, I think that could work.

The next issue is: how does the compiler know to use LLVM `const_struct` when there's `Int32.static_array` or `Int32.slice` which basically create the data "at runtime"? That's why I was thinking that having a literal for this is simpler, because you have the data right there. But if we figure out how to do it, then I think I like the idea.

**Comment by HertzDevil** [comment]
> Flipping the idea around, is there ever a time when an author would set a constant and *want* it to be mutable? If no, then do we even need the annotation?

`Random::DEFAULT` is obviously mutable, and the Crystal standard library is not opinionated enough to expose this in a class method.

**Comment by Fryguy** [comment]
Perhaps that's where the annotation might be helpful in declaring the exceptions to the rule?  I could see Random::DEFAULT having a `[@Mutable]` annotation, since it's an outlier.

It is strange that one is a constant - it's really more of a singleton instance.

**Comment by asterite** [comment]
> it's really more of a singleton instance

Yes, that's what constants are in Ruby and Crystal. They are called constants, but they are just write-once things. They can be mutable, and that's totally fine and expected (this fact is used a lot! we can't change this)

**Comment by HertzDevil** [comment]
A brief plan is as follows. The read-only constants would most likely be intercepted here:

https://github.com/crystal-lang/crystal/blob/0ca193126ecc080566367af2ba9cfdab3f81fca6/src/compiler/crystal/semantic/top_level_visitor.cr#L794-L802

If `@[ReadOnly]` is present, we set a flag in `const` so that we could later do the constant generation logic. This is lazily done like regular constants, so the initializer expression is not visited unless requested, and `Int32.static_array` would not expand immediately despite being a macro call. The other place is:

https://github.com/crystal-lang/crystal/blob/0ca193126ecc080566367af2ba9cfdab3f81fca6/src/compiler/crystal/semantic/main_visitor.cr#L180-L200

If `type` is read-only, we set a `needs_read_only_constant?` flag on the nested `MainVisitor`, inspired by `needs_type_filters?`. This allows us to avoid expanding the `static_array` macro call altogether:

```crystal
def visit(node : Call)
  prepare_call(node)

  if needs_read_only_constant?
    obj = node.obj
    if obj && node.name == "static_array"
      element_type = (path_lookup || scope).lookup_type?(obj)
      obj.raise "..." unless element_type && element_type.constant_type?
      node.args.each &.accept(self)
      @read_only_constant = ...
      node.type = @program.static_array_of(element_type, node.args.size)
      return false
    end

    node.raise "..."
  end

  if expand_macro(node)
    # ...
  end
  # ...
end
```

Alternatively we could have a separate visitor type, but reusing `MainVisitor` might reduce the friction involved when more constant expressions are supported. The `@read_only_constant` is then assigned to the `Const` node that triggers the visitation (possibly leveraging `Const#compile_time_value` which has pretty much the same purpose). This variable can simply be a fully typed clone of the whole initializer expression; in practice it might be structurally different if e.g. we allow the use of `MathInterpreter` while computing this read-only constant node.

That should give enough information to `CodeGenVisitor#initialize_simple_constants`, which then builds the constants recursively and registers them through `@llvm_mod.globals.add`. For `StaticArray` a `LibLLVM.const_array` is generated. For `Slice`, the elements are emplaced in read-only memory as a `const_array` too, and the slice points at it with `@read_only` set to true. (This is the one part I am least confident about, so I hope this inner pointer between constants is supported.)

**Comment by HertzDevil** [comment]
> The main issue is that `Slice` and `StaticArray` are mutable types. So how does the compiler know to prevent modifying them? For example:
> 
> ```crystal
> @[ReadOnly]
> DATA = Int32.static_array(1, 2, 3)
> 
> data = DATA
> data[0] = 1 # How do we detect this? We have to somehow carry the "readonly" information
> 
> data2 = StaticArray(Int32, 3).new(0)
> data3 = rand < 0.5 ? data : data2
> data3[0] # How do we know this is allowed? What's the type of data3?
> ```

This is actually a very valid concern if `DATA` were a reference instead of a value. If that were the case, `data` must be read-only, and so is `data3`, analogous to `const &` propagation in C++. I believe const-qualified and mut-qualified types are counterproductive to Ruby and Crystal, where the usual response to situations like this is to exercise developer discipline. This is actually one of the reasons why I limited the first iteration to selected aggregates of value types.

In the opposite direction, if mutability checks are deferred to runtime, _any_ type can implement its own `read_only?` flag similar to `Slice`. This, of course, is known in Ruby as `frozen?`.

**Comment by asterite** [comment]
>  I believe const-qualified and mut-qualified types are counterproductive to Ruby and Crystal, where the usual response to situations like this is to exercise developer discipline

I totally agree with this! I think `const` and `immutable` look nice, but I don't consider them "fun".

I think for being truly optimal, Crystal needs a way to specify arrays or aggregates stored in read-only memory, that LLVM will aggressively inline. So just being able to do that in the least annoying way would work for me.

I'm not sure hardcoding code if the call name is "static_array" is a good idea. Maybe the method could have a compiler-only annotation similar to `@[Primitive]` where the compiler knows about it and knows what to do about it, like perform certain optimizations. That means that if a user redefines `static_array` to be something else, we won't produce potentially invalid code.

Of course, feel free to experiment with all of this!

**Comment by j8r** [comment]
Beware that mutable constants are used as a hack to share a common read-write variable across macro scopes.
